File: 502. IPO.py
# 10 min write code, 20min debug
# TLE: q.get() will block util the q has element.......
# check the size then pass.
import queue
class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        n = len(profits)
        capAndProfit = zip(capital, profits)
        sortedCapAndProfit = sorted(capAndProfit, key=lambda x: (x[0], x[1]))
        print(sortedCapAndProfit)
        nextEvent = 0
        q = queue.PriorityQueue()
        for kk in range(k):
            while nextEvent < n and w >= sortedCapAndProfit[nextEvent][0]:
                event = sortedCapAndProfit[nextEvent]
                q.put((-event[1], event[1]))
                nextEvent += 1
            if not q.empty():
                cur = q.get()
                w += cur[1]
        return w

        

File: 502. IPO.cpp
/*
use cpp to write this again.
*/

class Solution {
private:
    struct Node {
        int profit;
        int capital;
        Node(int p, int c) {
            profit = p;
            capital = c;
        };
    };
    struct MinProfit {
        int profit;
        MinProfit(int p) { profit = p; };
        bool operator<(const MinProfit& rhs) const {
            return profit < rhs.profit;
        };
    };
    static int cmp(const Node& p, const Node& q) {
        if (p.capital != q.capital)
            return p.capital < q.capital;
        return q.profit < p.profit;
    }

public:
    int findMaximizedCapital(int k, int w, vector<int>& profits,
                             vector<int>& capital) {
        vector<Node> cnp;
        for (int i = 0; i < profits.size(); i++) {
            cnp.emplace_back(profits[i], capital[i]);
        }
        sort(cnp.begin(), cnp.end(), cmp);
        int nextEvent = 0;
        priority_queue<MinProfit> q;
        for (int kk = 0; kk < k; kk++) {
            while (nextEvent < cnp.size() && w >= cnp[nextEvent].capital) {
                MinProfit mp = MinProfit(cnp[nextEvent].profit);
                q.push(mp);
                nextEvent++;
            }
            if (!q.empty()) {
                MinProfit mp = q.top();
                q.pop();
                w += mp.profit;
            }
        }
        return w;
    }
};